### ISSUE TITLE
Improve `add_signature_parameters` : add `first_possible` and `last_possible`

### ISSUE THREAD

**Issue by smarie** [smarie/python-makefun#82]
Sometimes adding a parameter at the beginning or end of parameters list would make the resulting signature invalid. Typically, adding a keyword-only parameter before a keyword-or-positional, or adding a parameter without default after a parameter with a default (and conversely)...

Two new arguments `first_possible` and `last_possible` could allow the users to ask for a "smart insert first" and "smart insert last" feature.

This would help solving https://github.com/smarie/python-pytest-cases/issues/279
in a reusable way.

**Comment by vyasr** [comment]
Is there any scenario where you wouldn't want `last` to behave like `last_possible`? It seems to me like if we were to add these two new options rather than modifying the existing ones, then at minimum we would want `first` and `last` to throw errors if the resulting signatures would be ill-formed. At that point is there much benefit to not just making this behavior the default for `first` and `last`?

**Comment by vyasr** [comment]
On second thought, I am rethinking my position. Explicit is better than implicit after all. I think changing `first` and `last` to throw errors along with this change would be sensible, but automatically remapping the position may be a bit more magic than we want by default.

**Comment by vyasr** [comment]
How should this be handled in the case where parameters are of different kinds? It's fairly obvious how `last_possible` should behave when all the parameters are positional, for instance, but what happens if some are positional and some are keyword? Do the positional ones go at the end of the list of positional parameters, while the keyword ones go at the end of the kwargs list? Should these work with functions containing positional-only or keyword-only parameters?

**Comment by smarie** [comment]
> Do the positional ones go at the end of the list of positional parameters, while the keyword ones go at the end of the kwargs list? Should these work with functions containing positional-only or keyword-only parameters?

Yes, I guess that the "conctract" should be based on the most intuitive behaviour for users. So for example

 - `last_possible`: each parameter will be inserted at the right-most possible position in the signature, while preserving Syntactic validity. When possible, the initial order of provided parameters is preserved.

**Comment by vyasr** [comment]
I can implement this, but tbh I'm not sure how valuable this feature will be. I think we may be better off resolving https://github.com/smarie/python-pytest-cases/issues/279 directly in pytest-cases rather than trying to add this functionality. The problem that I see is that, except in very special cases like pytest tests where pytest is handling function parameters via its own dependency injection mechanisms, it is generally not very useful to modify a function signature without knowing the exact desired result a priori. Wouldn't it make more sense to put the onus on the calling code to figure out what the first or last valid position is? That way the calling code also actually knows what valid ways to call the function are. Otherwise, there's no guarantee that a user creating a function in this way will have enough information about the signature to use it correctly afterwards. For instance, they could easily pass parameters in the wrong order.

**Comment by lucaswiman** [comment]
I think adding a keyword-only argument with a default is almost always safe, unless there is an existing argument of the same name. Supporting that case seems useful. It also seems reasonable that makefun would figure out where it's safe to insert an argument and put it there.

**Comment by vyasr** [comment]
I agree, this could be made to work for a keyword-only parameter. My concern applies more to positional or positional-or-keyword parameters, where the calling code can't make assumptions about the resulting API so it doesn't seem very useful. Maybe we should implement this and just restrict this to keyword-only parameters, then. Does that sound reasonable to both of you @lucaswiman @smarie?

If we do go in that direction, this won't really address the original problem in https://github.com/smarie/python-pytest-cases/issues/279, but I think that's OK.

**Comment by vyasr** [comment]
No worries, my workaround is perfectly sufficient for now so I'm happy to try to resolve this first.

> Maybe one way to start could be to write a few tests first, to be sure we agree on what this should do ?

Yes, this is how I started and what made me ask all these questions. I am happy to put together an implementation early next week in a draft PR that we can iterate on.
